homeassistant/components/rabbitair/__init__.py

```diff
@@ -0,0 +1,83 @@
+"""The Rabbit Air integration."""
+from __future__ import annotations
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from rabbitair import Client, State, UdpClient
+
+from homeassistant.components import zeroconf
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_ACCESS_TOKEN,
+    CONF_HOST,
+    CONF_SCAN_INTERVAL,
+    Platform,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DEFAULT_SCAN_INTERVAL, DOMAIN, KEY_COORDINATOR, KEY_DEVICE
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.FAN]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Rabbit Air from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    host: str = entry.data[CONF_HOST]
+    token: str = entry.data[CONF_ACCESS_TOKEN]
+
+    _LOGGER.debug("Initializing with host %s, token %s", host, token[:5])
+    zeroconf_instance = await zeroconf.async_get_async_instance(hass)
+    device: Client = UdpClient(host, token, zeroconf=zeroconf_instance)
+
+    scan_interval: int = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    update_interval = timedelta(seconds=scan_interval)
+
+    async def async_update_data() -> State:
+        try:
+            return await device.get_state()
+        except asyncio.TimeoutError:
+            raise
+        except Exception as err:
+            raise UpdateFailed from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="rabbitair",
+        update_method=async_update_data,
+        update_interval=update_interval,
+    )
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = {
```

Comment:
We could make a NamedTuple to hold the coordinator and device. It would be easier to use and easier to maintain typing.

Reply:
Thanks for the tip! I made that change. But then, after rewriting the code related to the update coordinator, it became sufficient to pass only the coordinator itself, so I removed the NamedTuple.